content/en/docs/concepts/architecture/controller.md

```diff
@@ -26,7 +26,7 @@ closer to the desired state, by turning equipment on or off.
 ## Controller pattern
 
 A controller tracks at least one Kubernetes resource type.
-These [objects](/docs/concepts/overview/working-with-objects/kubernetes-objects/#kubernetes-objects)
+These [objects]({{< ref "/docs/concepts/overview/working-with-objects/kubernetes-objects.md#kubernetes-objects" >}})
```

Comment:
I'm wondering if this is literally should be the absolute path of the link. Also another nit: Probably we can give the best practices of writing links somewhere so others can adopt this way of writing a link? wdyt

Reply:
Yes. The `ref` shortcode requires the markdown file name. So there are pros and cons:

The benefits as I see it:

- By explicitly using file names in the links, we can avoid dangling links. For every links written this way, Hugo can help ensure the target does exists, and it can do so for every language translation.
- As a side-effect, we can kill some link targets that are actually redirects. For example, `[kube-apiserver](/docs/admin/kube-apiserver/)` has to be rewritten into `[kube-apiserver]({{< ref "/docs/reference/command-line-tools-reference/kube-apiserver.md" >}})`.

The drawbacks is that rewriting the links is a tedious task because for each link we need to check:

1. if the link `[foo](/docs/bar/)` is there, whether we do have `[foo](/docs/bar.md)` in the file system. If not, we need to check,
2. if the target file `/docs/bar/_index.md` is there, and revise the link to `[foo](/docs/bar/_index.md)` if so. Or
3. if we have an entry in `/static/_redirects` that says `/docs/foo/  /docs/zoo/ 301`. We need to check `/docs/zoo/` by repeating step 1 and 2. Or else,
4. we need to check if the target is of pattern `/docs/reference/generated/kube*`. We have to leave the target as is in this case because the target is not meant to be localized. The target file lives in `/static/docs/...` directory.
5. finally, if the target itself contains a shortcode, we cannot change it. We have to leave it there. For example, `[latest version](/docs/path1/{{< latest-version >}}/path2)`.

Reply:
So I had given this some additional thoughts these days. Maybe I was too ambitious to start this thread. I'm not sure such a logic can find its home in GoHugo. Even if I can complete this in a week or so, how difficult it is for doc maintainers to check if all future link changes are obeying such a complicated rule, in order to be both valid and friendly to localization teams.

I'm myself kinda split on this. Maybe we should instead develop a script for this instead? Since we almost know all the gocha's out there, we can come up with a tool for all localization teams to use. Maybe such a tool can be combined with the link checking for English site as well? I don't know. I'm just dumping my brain here.

Reply:
I seconded the idea of automation and link checker for this. Since we know the pattern forming automation will be easier since we know the possible gotchas hence we can address this by forming rigorous testing. Need your input on this @sftim wdyt?

Reply:
I'm minded to look a little more to see if we can convert a hyperlink within a localization `eg` that appears in Markdown as (say) `/docs/tasks/tools/install-kubectl/` so that the rendered HTML has `<a href="/eg/docs/concepts/docs/tasks/tools/install-kubectl/>…</a>`. In other words, the site handles the conversion automatically.

If that's feasible, that's my favorite option. There might be a change to get advice on this from the people behind Hugo, too.

If we had the automatic conversion in place, we could run a one-time script to clean up all of the links that manually include the localization code.

Reply:
Do you know whom should we contact regarding this Hugo config?

Reply:
Hugo has its own page about [requesting help](https://discourse.gohugo.io/t/requesting-help/9132) so that might be one place to turn.

I'd be delighted to learn that some other group that uses Hugo has already found this problem and prepared a good solution.